app/models/concerns/foreman_resource_quota/host_managed_extensions.rb

```diff
@@ -12,50 +13,95 @@ module HostManagedExtensions
+      verify_resource_quota_limits(quota_utilization, host_resources)
+    end
+
+    def handle_error(error_module, error_message, log_message)
+      errors.add(error_module, error_message)
+      Rails.logger.error(N_(log_message))
+      false
+    end
 
-      resource_quota.determine_utilization([self])
-      unless resource_quota.missing_hosts.empty?
-        errors.add(:resource_quota,
-          N_(format('An error occured reading host resources. Check the foreman error log.')))
-        return false
+    def determine_quota_utilization
+      resource_quota.determine_utilization
+      missing_hosts = resource_quota.missing_hosts
+      unless missing_hosts.empty?
+        raise ResourceQuotaUtilizationException,
+          "Resource Quota '#{resource_quota.name}' cannot determine resources for #{missing_hosts.size} hosts."
       end
+      resource_quota.utilization
+    end
 
-      verify_resource_quota_limits(resource_quota.utilization)
-      errors.empty?
+    def determine_host_resources
+      (host_resources, missing_hosts) = call_utilization_helper(resource_quota.active_resources, [self])
+      unless missing_hosts.empty?
+        raise HostResourcesException,
+          "Cannot determine host resources for #{name}"
+      end
+      host_resources
     end
-    # rubocop: enable Metrics/AbcSize
 
-    private
+    def verify_resource_quota_limits(quota_utilization, host_resources)
+      quota_utilization.each do |resource_type, resource_utilization|
+        next if resource_utilization.nil?
+
+        max_quota = resource_quota[resource_type]
+        all_hosts_utilization = resource_utilization + host_resources[resource_type.to_sym]
+        next if all_hosts_utilization <= max_quota
 
-    def verify_resource_quota_limits(quota_utilization)
-      quota_utilization.each do |type, utilization|
-        next if utilization.nil?
-        max_quota = resource_quota[type]
-        next if utilization <= max_quota
-        errors.add(:resource_quota, N_(format("Host resources exceed quota '%s' for %s: %s > %s",
-          resource_quota_name,
-          natural_resource_name_by_type(type),
-          utilization,
-          max_quota)))
-        break
+        raise ResourceLimitException, formulate_limit_error(resource_utilization,
+          all_hosts_utilization, max_quota, resource_type)
       end
     end
 
+    def formulate_limit_error(resource_utilization, all_hosts_utilization, max_quota, resource_type)
+      error_text = if resource_utilization < max_quota
+                     "Host exceeds %s limit of '%s'-quota by %s (max. %s)"
+                   else
+                     "%s limit of '%s'-quota exceeded by %s (max. %s)"
+                   end
+      N_(format(error_text, natural_resource_name_by_type(resource_type),
+        resource_quota.name,
+        resource_value_to_string(all_hosts_utilization - max_quota, resource_type),
+        resource_value_to_string(max_quota, resource_type)))
```

Comment:
What are the different error cases?

Reply:
Changing to `"%s limit of '%s'-quota is already exceed by %s (max. %s)"`?
I want to keep the same order of the parameters, so I can keep the same `format` call for both strings :crossed_fingers:

Reply:
Maybe you could write something like `"%s limit of '%s'-quota is already exceed by %s without adding the new host (max. %s)"` ?

Reply:
With this wording, it means explicitly the difference between max_quota and its current utilization (without the host). Therefore, I've changed it to two different format.

IMHO its also better like this.